SDK/AppCenter/Microsoft.AppCenter.Windows.Shared/Storage/StorageAdapter.cs

```diff
@@ -1,158 +1,193 @@
+            }
+            return raw.sqlite3_finalize(stmt);
         }
 
-        private static StorageException ToStorageException(SQLiteException e)
+        private List<object[]> ExecuteSelectionSqlQuery(string query, IList<object> args = null)
         {
-            return new StorageException($"SQLite errorCode={e.Result}", e);
+            var db = _db ?? throw new StorageException("The database wasn't initialized.");
+            var entries = new List<object[]>();
+            var queryResult = raw.sqlite3_prepare_v2(db, query, out var stmt);
+            if (queryResult != raw.SQLITE_OK)
+            {
+                var errorMessage = raw.sqlite3_errmsg(_db);
+                throw new StorageException($"Failed to prepare SQL query, result={queryResult}\n\t{errorMessage}");
+            }
+            BindParameters(stmt, args);
+            while (raw.sqlite3_step(stmt) == raw.SQLITE_ROW)
+            {
+                var count = raw.sqlite3_column_count(stmt);
+                entries.Add(Enumerable.Range(0, count).Select(i => GetColumnValue(stmt, i)).ToArray());
+            }
+            var result = raw.sqlite3_finalize(stmt);
+            if (result != raw.SQLITE_OK)
+            {
+                var errorMessage = raw.sqlite3_errmsg(_db);
+                throw new StorageException($"Failed to finalize SQL query, result={result}\n\t{errorMessage}");
+            }
+            return entries;
         }
 
-        public async Task<int> DeleteAsync<T>(Expression<Func<T, bool>> pred) where T : new()
+        public void CreateTable(string tableName, string[] columnNames, string[] columnTypes)
         {
-            try
+            var tableClause = string.Join(",", Enumerable.Range(0, columnNames.Length).Select(i => $"{columnNames[i]} {columnTypes[i]}"));
+            var result = ExecuteNonSelectionSqlQuery($"CREATE TABLE IF NOT EXISTS {tableName} ({tableClause});");
+            if (result != raw.SQLITE_OK)
             {
-                var numDeleted = 0;
-                var table = _dbConnection.Table<T>();
-                var entries = await table.Where(pred).ToListAsync().ConfigureAwait(false);
-                foreach (var entry in entries)
-                {
-                    numDeleted += await _dbConnection.DeleteAsync(entry).ConfigureAwait(false);
-                }
-                return numDeleted;
+                var errorMessage = raw.sqlite3_errmsg(_db);
+                throw new StorageException($"Failed to create table, result={result}\n\t{errorMessage}");
             }
-            catch (SQLiteException e)
+        }
+
+        public int Count(string tableName, string columnName, object value)
+        {
+            var result = ExecuteSelectionSqlQuery($"SELECT COUNT(*) FROM {tableName} WHERE {columnName} = ?;", new[] { value });
+            return (int)(long)(result.FirstOrDefault()?.FirstOrDefault() ?? 0L);
```

Comment:
Two typecast in a row is not very readable. Could you break it down into two lines?